library/src/rng/distribution/discrete.hpp

```diff
@@ -100,15 +101,15 @@ class rocrand_discrete_distribution_base : public rocrand_discrete_distribution_
         {
             if (probability != NULL)
             {
-                ROCRAND_HIP_FATAL_ASSERT(hipFree(probability));
+                ROCRAND_HIP_FATAL_ASSERT(hipFreeAsync(probability, m_stream));
```

Comment:
I've managed to fix up some of the issues with the Poisson distribution, such the all tests in test/internal/test_poisson_distribution.cpp now pass with graphs. I'll push a new commit shortly.
I'll also leave in the failing test cases so that you can see them at runtime. If you have any insight into what might be causing them, I'd really appreciate any feedback you have.

Most of the generators' test_generate_poisson test cases seem to fail due to incorrect results, though it can be intermittent. I've played around with the placement of the stream capture block (capturing progressively narrower slices of the generator code), but I still haven't been able to nail down exactly what the problem is.

Reply:
Looking further into this, the problem is that the host data used in the `hipMemcpyAsync` in `create_alias_table` (and `create_cdf`) is already destroyed by the time the graph is launched. This should also cause a fail in the `TEST_P(rocrand_generate_poisson_tests, generate_test)` test. In my testing it does not fail, I'm not sure why, but the host memo

I'm not sure if this can be solved in a satisfying way. I think the best thing to do would be to make an exception for `rocrand_generate_poisson` after all. We could later extend the API to add an `rocrand_generate_discrete` that uses the `rocrand_discrete_distribution` constructed by `rocrand_create_poisson_distribution` or `rocrand_create_discrete_distribution`. It should then be safe to record and launch the `rocrand_generate_discrete` calls.

Reply:
Similar problems exist in e.g. `TEST_P(rocrand_kernel_xorwow_poisson, rocrand_discrete)`, where `rocrand_destroy_discrete_distribution` is captured in the stream, but this will destroy the host input for the copy. I think the test should look like:
```
rocrand_create_poisson_distribution

createGraphHelper
rocrand_discrete_kernel
endCaptureGraphHelper

rocrand_destroy_discrete_distribution

copy h2d and test
```

Reply:
Thanks @nolmoonen, I appreciate your help with tracking this problem down. I spoke with Colin a while back, and he'd mentioned it should be fine to make an exception for the poisson distribution like you suggested, just as long as we add a note about it to the changelog. Things are quite busy right now, but I'll do my best to have that up shortly.